contracts/CumulativeTwap.sol

```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: MIT License
+pragma solidity 0.7.6;
+pragma experimental ABIEncoderV2;
+
+import { BlockContext } from "./base/BlockContext.sol";
+
+contract CumulativeTwap is BlockContext {
+    //
+    // STRUCT
+    //
+    struct Observation {
+        uint256 price;
+        uint256 priceCumulative;
+        uint256 timestamp;
+    }
+
+    //
+    // EVENT
+    //
+    event PriceUpdated(uint256 price, uint256 timestamp, uint8 indexAt);
+
+    //
+    // STATE
+    //
+    // let's use 15 mins and 1 hr twap as example
+    // if the price is being updated 15 secs, then needs 60 and 240 historical data for 15mins and 1hr twap.
+    Observation[256] public observations;
+
+    uint8 public currentObservationIndex;
+
+    function _update(uint256 price, uint256 lastUpdatedTimestamp) internal {
+        // for the first time update
+        if (currentObservationIndex == 0 && observations[0].timestamp == 0) {
+            observations[0] = Observation({ price: price, priceCumulative: 0, timestamp: lastUpdatedTimestamp });
+            emit PriceUpdated(price, lastUpdatedTimestamp, 0);
+            return;
+        }
+
+        // CT_IT: invalid timestamp
+        // add `==` in the require statement in case that two or more price with the same timestamp
+        // this might happen on Optimism bcs their timestamp is not up-to-date
+        Observation memory lastObservation = observations[currentObservationIndex];
+        require(lastUpdatedTimestamp >= lastObservation.timestamp, "CT_IT");
+
+        // overflow of currentObservationIndex is desired since currentObservationIndex is uint8 (0 - 255),
+        // so 255 + 1 will be 0
+        currentObservationIndex++;
+
+        uint256 elapsedTime = lastUpdatedTimestamp - lastObservation.timestamp;
+        observations[currentObservationIndex] = Observation({
+            priceCumulative: lastObservation.priceCumulative + (lastObservation.price * elapsedTime),
+            timestamp: lastUpdatedTimestamp,
+            price: price
+        });
+
+        emit PriceUpdated(price, lastUpdatedTimestamp, currentObservationIndex);
+    }
+
+    function _getPrice(
+        uint256 interval,
+        uint256 latestPrice,
+        uint256 latestUpdatedTimestamp
+    ) internal view returns (uint256) {
+        Observation memory lastestObservation = observations[currentObservationIndex];
+        if (lastestObservation.price == 0) {
+            // CT_ND: no data
+            revert("CT_ND");
+        }
+
+        uint256 currentTimestamp = _blockTimestamp();
+        uint256 targetTimestamp = currentTimestamp - interval;
```

Comment:
Should we use SafeMath here?

Reply:
fixed, forgot why we didn't use SafeMath in the first place, applied SafeMath now.